File: src/ariadne_sdk/instrument/anthropic.py
"""Anthropic 自动埋点。

与 OpenAI 版同构：显式包装 messages.create 这一公开方法。

用法：
    from anthropic import Anthropic
    from ariadne_sdk.instrument.anthropic import wrap_anthropic

    client = wrap_anthropic(Anthropic())
"""

from __future__ import annotations

import contextlib
import functools
from typing import Any, TypeVar

from ariadne_sdk.client import get_client
from ariadne_sdk.instrument.common import (
    extract_anthropic_usage,
    preview_messages,
    preview_output,
)

T = TypeVar("T")

_PROVIDER = "anthropic"


def _instrument_create(original: Any) -> Any:
    @functools.wraps(original)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        client = get_client()
        if client is None:
            return original(*args, **kwargs)

        model = str(kwargs.get("model", ""))
        span = client.span(
            f"chat {model}".strip(),
            kind="llm",
            operation="chat",
            provider=_PROVIDER,
            model=model,
        )
        with span:
            span.set_attributes(
                {
                    "temperature": kwargs.get("temperature"),
                    "max_tokens": kwargs.get("max_tokens"),
                    "stream": kwargs.get("stream", False),
                    # system prompt 单独记长度，便于判断上下文构成
                    "system_len": len(str(kwargs.get("system", ""))) or None,
                }
            )
            if messages := kwargs.get("messages"):
                span.set_input(preview_messages(messages))

            response = original(*args, **kwargs)

            if kwargs.get("stream"):
                span.set_attributes({"usage_unavailable": "stream"})
                return response

            span.set_usage(**extract_anthropic_usage(response))
            actual_model = getattr(response, "model", "") or model
            span.set_model(model, response_model=str(actual_model), provider=_PROVIDER)
            span.set_output(preview_output(response))
            if stop_reason := getattr(response, "stop_reason", ""):
                span.set_attributes({"finish_reason": str(stop_reason)})
            return response

    return wrapper


def wrap_anthropic(client: T) -> T:
    """就地包装 Anthropic 客户端。幂等。"""
    messages = getattr(client, "messages", None)
    if messages is None:
        return client

    create = getattr(messages, "create", None)
    if create is None or getattr(create, "_ariadne_wrapped", False):
        return client

    wrapped = _instrument_create(create)
    wrapped._ariadne_wrapped = True
    # 某些版本用 __slots__ 或只读属性，跳过而非崩掉业务
    with contextlib.suppress(AttributeError, TypeError):
        messages.create = wrapped
    return client

File: sdk/typescript/src/instrument/anthropic.ts
/**
 * Anthropic 自动埋点。
 *
 * 与 Python SDK 对等：包装 Anthropic SDK 的 messages.create 方法。
 *
 * 用法：
 *   import Anthropic from "@anthropic-ai/sdk";
 *   import { wrapAnthropic } from "@ariadne/sdk/instrument/anthropic";
 *   const client = wrapAnthropic(new Anthropic({ apiKey: "..." }));
 *   await client.messages.create({ model: "claude-3-opus", messages: [...] });
 */

import { getClient } from "../client";

const PROVIDER = "anthropic";

interface AnthropicRequest {
  model?: string;
  messages?: unknown;
  system?: string;
  max_tokens?: number;
  temperature?: number;
  stream?: boolean;
}

interface AnthropicResponse {
  model?: string;
  usage?: {
    input_tokens?: number;
    output_tokens?: number;
    cache_read_input_tokens?: number;
    cache_creation_input_tokens?: number;
  };
  content?: Array<{ text?: string }>;
}

function previewMessages(messages: unknown): string {
  if (!Array.isArray(messages)) return "";
  return messages
    .slice(0, 5)
    .map((m: { role?: string; content?: unknown }) => {
      const content = typeof m.content === "string" ? m.content : JSON.stringify(m.content);
      return `${m.role ?? "?"}: ${content.slice(0, 200)}`;
    })
    .join("\n")
    .slice(0, 1024);
}

function previewOutput(resp: AnthropicResponse): string {
  if (resp.content?.[0]?.text) return resp.content[0].text.slice(0, 1024);
  return "";
}

export function wrapAnthropic<T extends Record<string, unknown>>(client: T): T {
  const messages = (client as unknown as {
    messages?: { create?: (...a: unknown[]) => unknown };
  }).messages;

  if (messages && typeof messages.create === "function") {
    const original = messages.create;
    messages.create = function (...args: unknown[]): unknown {
      const request = (args[0] as AnthropicRequest) ?? {};
      const ariadne = getClient();
      if (!ariadne) return original.apply(messages, args);

      const model = request.model ?? "";
      const span = ariadne.span(`messages.create ${model}`.trim(), {
        kind: "llm",
        operation: "messages.create",
        provider: PROVIDER,
        model,
      });

      span.setAttributes({
        temperature: request.temperature,
        max_tokens: request.max_tokens,
        stream: request.stream ?? false,
      });

      if (request.messages) span.setInput(previewMessages(request.messages));
      if (request.system) {
        span.setAttributes({ system_prompt_length: request.system.length });
      }

      span.enter();

      const result = original.apply(messages, args);

      if (result instanceof Promise) {
        return result.then(
          (resp: unknown) => {
            const r = resp as AnthropicResponse;
            if (r?.model) span.setModel(model, { responseModel: r.model });
            if (r?.usage) {
              span.setUsage({
                input_tokens: r.usage.input_tokens ?? 0,
                output_tokens: r.usage.output_tokens ?? 0,
                cache_read_tokens: r.usage.cache_read_input_tokens ?? 0,
                cache_write_tokens: r.usage.cache_creation_input_tokens ?? 0,
                reasoning_tokens: 0,
              });
            }
            span.setOutput(previewOutput(r));
            span.end();
            return resp;
          },
          (err: Error) => {
            span.recordError(err);
            span.end();
            throw err;
          },
        );
      }

      const r = result as AnthropicResponse;
      if (r?.model) span.setModel(model, { responseModel: r.model });
      if (r?.usage) {
        span.setUsage({
          input_tokens: r.usage.input_tokens ?? 0,
          output_tokens: r.usage.output_tokens ?? 0,
          cache_read_tokens: r.usage.cache_read_input_tokens ?? 0,
          cache_write_tokens: r.usage.cache_creation_input_tokens ?? 0,
          reasoning_tokens: 0,
        });
      }
      span.setOutput(previewOutput(r));
      span.end();
      return result;
    };
  }

  return client;
}
